client/packages/core/src/queryTypes.ts

```diff
@@ -70,9 +180,36 @@ type Direction = 'asc' | 'desc';
 
 type Order = { [key: string]: Direction };
 
-type $Option<Fields extends string[]> = {
+type $Option<
+  Fields extends string[],
+  WhereFieldTypes extends {
+    [key: string]: any;
+  } = Record<any, BSUnknown>,
+> = {
+  $?: {
+    where?: WhereClause<WhereFieldTypes>;
+    order?: Order;
+    limit?: number;
+    last?: number;
+    first?: number;
+    offset?: number;
+    after?: Cursor;
+    before?: Cursor;
+    fields?: Fields;
+  };
+};
+
+// Typed version that supports dot notation for nested queries
+type $OptionWNest<
+  S extends IContainEntitiesAndLinks<any, any>,
+  K extends keyof S['entities'],
+  Fields extends string[],
```

Comment:
nit: but I wonder if we need to actually pass these in as third and forth args. They can both be derived from `S` and `K`. It might make this type simpler to read